cogent/src/Cogent/DataLayout/Desugar.hs

```diff
@@ -1,27 +1,73 @@
-                                  , RepExpr(RepRef, Prim, Offset, Record, Variant)
-                                  )
+import Cogent.Compiler              ( __fixme
+                                    , __impossible
+                                    )
+import Cogent.Common.Syntax         ( DataLayoutName
+                                    , Size
+                                    , TagName
+                                    , FieldName
+                                    )
+import Cogent.Common.Types          ( Sigil(Unboxed, Boxed))
+import Cogent.DataLayout.Surface    ( DataLayoutExpr
+                                    , DataLayoutSize
+                                    , RepSize(Bytes, Bits, Add)
+                                    , RepExpr(RepRef, Prim, Offset, Record, Variant)
+                                    )
+import Cogent.DataLayout.TypeCheck  (desugarSize)
 import Cogent.DataLayout.Core
+import Cogent.Core                  ( Type (TUnit, TPrim, TSum, TRecord, TCon, TVar, TVarBang, TFun, TString, TArray, TProduct)
+                                    )
+
+{- * DESUGARING 'Sigil's -}
+
+-- | After normalisation, 'TCon _ _ _' values only represent primitive and abstract types.
+--   Primitive types have no sigil, and abstract types may be boxed or unboxed but have no layout.
+--   'desugarAbstractTypeSigil' should only be used when desugaring the sigils of abstract types, to eliminate the `Maybe DataLayoutExpr`.
+desugarAbstractTypeSigil
+  :: Sigil (Maybe DataLayoutExpr)
+  -> Sigil ()
+desugarAbstractTypeSigil = fmap desugarMaybeLayout
+  where
+    desugarMaybeLayout Nothing = ()
+    desugarMaybeLayout _       = __impossible $ "desugarAbstractTypeSigil (Called on TCon after normalisation, only for case when it is an abstract type)"
+
+
+-- | If a 'DataLayoutExpr' was provided, desugars the 'DataLayoutExpr' to a 'DataLayout BitRange'
+--   Otherwise, constructs a 'DataLayout BitRange' which matches the type.
+--   TODO: Layout polymorphism and layout inference will require changing the second behavior /mdimeglio
+--
+--   Should not be used to desugar sigils associated with 'TCon _ _ _' types, ie. abstract types.
+desugarSigil
 
--- TODO: Split dataLayoutSurfaceToCore in Cogent.DataLayout.Typecheck
--- into a type checking function and this desugar function!
-desugarSize :: DataLayoutSize -> Size
-desugarSize (Bytes b) = b * 8
-desugarSize (Bits b)  = b
-desugarSize (Add a b) = desugarSize a + desugarSize b
+      -- | This type should be obtained by
+      --   1. Start with the raw type that the `Sigil` is attached to
+      --   2. Replace the top level sigil with `Unboxed`
+      --   3. Desugar the resulting type
+  :: (Type t)
+
+      -- | Since desugarSigil is only called for normalising boxed records (and later, boxed variants),
+      --   the 'Maybe DataLayoutExpr' should always be in the 'Just layout' alternative.
+  -> Sigil (Maybe DataLayoutExpr)
+  -> Sigil (DataLayout BitRange)
```

Comment:
Is it easier and cleaner to implement it as `desugarSigil :: Type t -> Sigil (DataLayout BitRange)` and pattern match on the sigil in each kind of type? IT was a little confusing why you need to break up a type into a type without top-level sigil and the sigil.

Reply:
Fair. Maybe it's worth making a comment.